R/shoji-catalog.R

```diff
@@ -240,28 +240,109 @@ setMethod("emails", "ShojiCatalog", function (x) getIndexSlot(x, "email"))
+        return(data.frame())
+    } else {
+        ### The following code is equivalent to out <- purrr::map_df(index, entry_to_df)
+        ################
+        entry_list <- lapply(index, entry_to_df, list_col_names = list_columns)
+        names   <- unique(unlist( lapply(entry_list, names)))
+        out <- data.frame(matrix(nrow = length(entry_list), ncol = length(names)))
+        names(out) <- names
+
+        for (i in seq_along(entry_list)) {
+            for (j in  names(entry_list[[i]])) {
+                out[i, j] <- entry_list[[i]][1, j]
+            }
+        }
+        #################
+
+        out <- as.data.frame(out, rownames = rownames, ...)
+
+        #When a bad key argument is passed to the the shoji index it returns a dataframe
+        # with an NA value. This exclude those columns.
+        exclude_cols <- grepl("^NA", names(out)) &
+            vapply(out, function(x)all(is.na(x)), FUN.VALUE = logical(1))
+        out <- out[, !exclude_cols, drop = FALSE]
+        #reorder columns to match the order in which keys were supplied
+        if (keys == TRUE) {
+            ordered_names <- names(out)
+        } else {
+            ordered_names <- names(out)[names(out) %in% keys]
+        }
+        out <- out[, ordered_names, drop = FALSE]
+
+        missing_keys <- dQuote(keys[!(keys %in% names(out))])
+
+        if (any(!(keys %in% names(out))) && keys != TRUE) {
+            if (length(missing_keys) == 1) {
+                error_text <-  " is an invalid key for catalogs of class "
+            } else {
+                error_text <- " are invalid keys for catalogs of class "
+            }
+            halt(paste0(
+                paste(missing_keys, collapse = ", ")),
+                error_text,
+                class(x),
+                ".")
+        }
+
+        return(out)
     }
-    return(out)
 }
+
+#' entry_to_df
+#' Turns an entry in the catalog into
+#'
+#' @param entry
+#' A single entry in a catalog
+#'
+#' @param list_col_names
+#' Some entries in a catalog do not fit neatly into a traditional dataframe and are
+#' instead stored in a list-column. This is controlled by passing the attribute names
+#' as the `list_col_names` argument. `
+#'
+#' @return A one row data frame
+#'
+entry_to_df <- function(entry, list_col_names ){
+            entry[vapply(entry, is.null, logical(1))] <- NA
+            vect_col <- entry[!(names(entry) %in% list_col_names)]
+            entry_df <- as.data.frame(vect_col, stringsAsFactors = FALSE)
+
+            if (any(names(entry) %in% list_col_names)) {
+                list_col <- entry[list_col_names]
+                list_df <- data.frame(matrix(nrow = 1, ncol = length(list_col)))
+                names(list_df) <- names(list_col)
+                for (i  in seq_along(list_col)) {
+                    list_df[[1, i]] <- list_col[i]
+                }
+                entry_df <- cbind(entry_df, list_df)
+            }
+            entry_df
+        }
```

Comment:
That said, I can test the other way and throw an error when something looks like a list column but isn't identified as one. Does that seem like a good solution?